Add legacy combo fire effect

cl8n opened the pull request:
- [x] https://github.com/ppy/osu-resources/pull/376

part of my [old gameplay features](https://github.com/ppy/osu/discussions/20659#discussioncomment-14002041) list

apart from the intentional changes I noted in the class docs, this is a pixel-perfect re-implementation of combo fire. much pain attempting to reverse engineer and debug this thing :melting_face: was rly cool to learn about everything involved though. the one thing I wasn't able to get working is hiding the fire while a spinner is active, I'm not sure what would be a sane way to get this component interacting with a DrawableSpinner and didn't find any similar examples

I also added `StoryFireInFront` support back for any old maps that may have used it

it can be added as a gameplay hud component in the skin editor. the fire container's height is scaled 0.25x, and the main fire piece 4x, so that the skin editor blueprint looks more reasonable instead of covering the entire editor. it's a little awkward for the fire to extend past the blueprint, but this is probably fine because it wouldn't make sense to use the top edge for positioning regardless.

cl8n committed:
Add combo fire and skin option

cl8n committed:
Add test scene

cl8n committed:
Respect StoryFireInFront beatmap setting

tadatomix commented:
On paper sounds interesting, but may I ask you either about providing a screenshot/video on how it looks like or tell how try it in game? Tried couple of ways, by looking at the code and can't say that it worked for me

peppy commented:
I was hoping for a reimagining with a better effect, but I guess this is something (have not checked out code yet).

If it's to be considered for merging, it must be global config and disabled by default though. 100% of people are going to want to disable this (rounding up).

I'd only consider making it on by default if we implemented a new version.

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -435,9 +435,14 @@ private Drawable createUnderlayComponents(WorkingBeatmap working)
                 RelativeSizeAxes = Axes.Both,
                 Children = new Drawable[]
                 {
+                    new LegacyComboFire(ScoreProcessor)
```
legacy elements cannot be in the player class. this should be skihnnable and the default should be "none" for non-legacy skins until a better visualisation is envisaged.

cl8n commented:
> On paper sounds interesting, but may I ask you either about providing a screenshot/video on how it looks like or tell how try it in game? Tried couple of ways, by looking at the code and can't say that it worked for me. Thank you in advance

I'll attach a video next time I work on it. to test in game, make sure you have the osu-resources branch checked out, have ran UseLocalResources, and are using a skin that either has `ComboFire: 1` or `Version:` < 2

> I was hoping for a reimagining with a better effect

that would be awesome for a future thing, I just wanted this old style to be available in one way or another as well. as you suggested I'll convert this into a properly skinnable component so that non-legacy skins can implement a more fitting effect

> If it's to be considered for merging, it must be global config and disabled by default though. 100% of people are going to want to disable this (rounding up).
>
> I'd only consider making it on by default if we implemented a new version.

I'm not sure if you saw the relevant part or not but it already is default disabled in most cases, and enabled by a skin.ini option. regardless I will have to revisit how the config works after the refactor mentioned above bc it obviously can't just be a skin.ini option if non-legacy skins implement it too

cl8n committed:
Merge master

cl8n committed:
Make combo fire a HUD/skinnable component

- Move to HUD namespace and implement ISerialisableDrawable
- Delete skin.ini stuff
- Scale the skin editor blueprint by 0.25x by moving the main fire
  effect to an internal drawable that applies a reciprocal scale
- Fix comboAlpha and break time alpha not applying invalidations
  correctly
- Use PositionAndColour vertex shader instead of custom one
- More comments on the alpha fields

cl8n committed:
Fix test scene

cl8n committed:
Add combo fire to default skin components if version < 2

cl8n committed:
Fix unused field warning

cl8n committed:
Fix and improve sizing for combo fire toolbox component

cl8n committed:
Fix bad initial state when reloading skin with player at high combo

cl8n committed:
Improve sizing for combo fire toolbox component... more

peppy committed:
Update resources
